src/dot.ts

```diff
@@ -41,7 +41,7 @@ export class Dot extends Modifier {
         const index = dot.checkIndex();
         props = note.getKeyProps()[index];
         // consider right displaced head with no previous modifier
-        if (right_shift === 0) shift = note.getRightDisplacedHeadPx();
+        if (note.checkModifierContext().getMembers('Parenthesis').length === 0) shift = note.getRightDisplacedHeadPx();
```

Comment:
This is a bit too specific, ideally `Dot` should not know about `Parenthesis`. Can we generalize somehow?

Reply:
I have not figure out a way. Previously Dot did not care, it made special calculations on the basis that Dot was always the first Modifier. Now Parenthesis can be first and this has to be checked for.

Reply:
I think this also masks the case where there is a displaced head with a parenthesis, no?

I'd suggest adding some kind of logic to `StaveNote`, where it can check its own members and return a shift. So only the note knows about whether or not there's a `Parenthesis`.

Reply:
Hi Mohit, I missed this one. I have been trying different approaches:
- Each element takes care of their shifts. I mean no extra code on dots to get closer to the note. I believe that this is the cleanest approach but then on the example above "Dot Multi_Voice Bravura" all dots are vertically aligned to the right of both notes. Anyway I believe that this example, as it is today, in the fourth group of notes if the notes above had two dots (the second one would overlap with the head of the note below).
- Then I tried to create a ModifierContext for each of the notes, and this works for the dots, but there are other tests that get overlap on the accidentals, so I discarded it.

It is quite a complicated exercise. 

My question would ti be acceptable to get display all the dots to the right of the notes aligned to the left of the head located in the right most position?

Reply:
The images related to the comment above:

**Current (as in this PR)**
![image](https://user-images.githubusercontent.com/22865285/150411653-6fb92bfc-c8d2-4695-a4d8-ee54c3ce6d8c.png)
**Each element taking care of its own shift. There is something to solve in the 4th group but the question is are the first and the second group ok drawn in this way?**
![image](https://user-images.githubusercontent.com/22865285/150411748-13cdab16-bbbf-4583-8c69-89aa4a5e9b06.png)

Reply:
@0xfe @ronyeh @mscuthbert @AaronDavidNewman @sschmidTU should I follow this way? I mean the approach from MuseScore.

Reply:
I am far from an expert in how engraved music should appear.

However, I do like the pictures where the dots are lined up with each other. It's very pleasing to the eye.

Reply:
The MuseScore arrangement looks visually beautiful, but it's incorrect by Gould "Behind Bars":

<img width="589" alt="Screen Shot 2022-01-25 at 08 38 40" src="https://user-images.githubusercontent.com/3521479/151038359-c98b46cf-c47e-4663-a6ec-a5b234a68aef.png">

That is for Chords with adjacent notes (clusters/7ths etc.).  When they are further apart or have no overlapping notes, then aligning the dots is preferred:

<img width="684" alt="Screen Shot 2022-01-25 at 08 40 53" src="https://user-images.githubusercontent.com/3521479/151038649-f521255c-869a-49f3-97a4-d979160d1cf7.png">


No automatic algorithm I know of, however, will get the 4th upstemmed group's dots to work properly with the third downstemmed group where there are two notes on the same line (D5) with different numbers of dots AND there isn't room for the downstem D5 to move its dots below (because of the C5 dots).  What an engraver would do in that case is have the dots for the first group appear before the second group's notes appear and move all dots closer to the notes than would be normal:

<img width="676" alt="Screen Shot 2022-01-25 at 08 41 04" src="https://user-images.githubusercontent.com/3521479/151038674-066825db-2be7-47df-966d-c01250272dae.png">

The top part may also be moved to the right (contrary to norms of alignment) if it clarifies the dot:

<img width="668" alt="Screen Shot 2022-01-25 at 08 42 43" src="https://user-images.githubusercontent.com/3521479/151038872-c59896cb-d4ca-49eb-83d8-f575994d4c73.png">

Once we get to overlapping parts, the rules are even more complex.

<img width="682" alt="Screen Shot 2022-01-25 at 08 43 04" src="https://user-images.githubusercontent.com/3521479/151038955-e683d373-7df3-49a7-a6da-c726ceebd0b8.png">

I would NOT expect Vexflow to take care of these situations at this point, but eventually (not for 4.0!) it'd be good to have utility functions to shift different voices at the same tick in different ways.  But I'm not volunteering to do this at this point.

Reply:
Thanks a lot @mscuthbert. What should happen with accidentals in conbination with these dots. I refer, for example, to the last stave above. We have note1, its associated dot, note2. What would happen if note 2 had a #?

Reply:
Say the accidental was on C (C#)  -- if the other voice doesn't have a C# then there would be little ambiguity (just ugly) if the sharp were before both groups.  If both groups had a C# then it's even better (even if there is some ambiguity).  If one group had C-natural and the other C# -- then you get into one of those really ugly situations where probably best would be to get out your rarely used squashed # sign that you carved from a wood block for just this occasion, and try to squeeze it in there.  :-D    -- I do have a friend who is a professional engraver whom I could ask about this case if you think we're actually going to support it now.  I'm pretty sure none of the professional software packages do this really.  (I still don't own Dorico though).

Reply:
Wow I certainly learned something from this thread. Yeah I think VexFlow should just try our best to accommodate the most common use cases, and then allow the user to add a relative shiftX / shiftY to any modifier, which will be applied after regular formatting. That way, a VexFlow user can just eyeball their score and tune by hand. 

This might not work if we are loading music xml and translating it to easy score for automated score creation. However, if we had an API that could say something like, in measure M, voice V, note group N, note index i, addRelativeShift(x, y) to the modifiers, that would allow us to "patch" rendered scores. Then a VexFlow user would have a patch for each musicxml file that they hand tweaked.

Reply:
@0xfe comment addressed. Doing the change I also fixed a bug on `Parenthesis`. The shifts in format were being added rather than calculating the biggest shift.

**Parenthesised Bravura**
![EasyScore Draw_Parenthesised Bravura](https://user-images.githubusercontent.com/22865285/151652584-22f6d1ce-15f5-4a47-b070-56e1d83a8d16.png)
**Current**
![EasyScore Draw_Parenthesised Bravura_current](https://user-images.githubusercontent.com/22865285/151652589-97e647e2-e9fe-4c1e-8118-16903d0cc2cd.png)
**Reference**
![EasyScore Draw_Parenthesised Bravura_reference](https://user-images.githubusercontent.com/22865285/151652593-bdc5f07e-190c-490a-99fb-b4334a91abf4.png)